src/core/message.c

```diff
@@ -24,11 +24,13 @@ typedef struct {
 
 // Underlying message structure.
 struct nng_msg {
-	uint32_t       m_header_buf[(NNI_MAX_MAX_TTL + 1)];
-	size_t         m_header_len;
-	nni_chunk      m_body;
-	uint32_t       m_pipe; // set on receive
-	nni_atomic_int m_refcnt;
+	uint32_t           m_header_buf[(NNI_MAX_MAX_TTL + 1)];
+	size_t             m_header_len;
+	nni_chunk          m_body;
```

Comment:
I'm not sure I have a better solution at present, but a few things about this design:

1. It increases the size of every message by 2 pointers (often this will be 16 bytes).
2. I think having protocol details in this particular manner may make moving messages between protocols very difficult.  I'm particular concerned about user applications that might want to proxy between e.g. SP PUB/SUB and MQTT.  I'd like to think about ways to make this better.  I need to research how you're using this before I can make a concrete suggestion.

Reply:
We did not touch any of the original parts of nng_msg, just follow your design in MQTT branch.

Reply:
And I am really interested on proxy between pub/sub of SP and MQTT, could you give me more detail on this? Maybe I can implement this feature at first.

Reply:
It would be pretty easy to take messages from an SP SUB and forward them to MQTT as publish messages.  It would be also easy to take incoming MQTT messages and turn them into SP PUB.  Of course QoS and some details like topic are going to get lost or need to be handled specially.

We should try to avoid burdening messages in a way that impedes this kind of usage.  If the message needs to convey protocol specific data (outside of the headers which wind up being "hidden" from applications when using "cooked" sockets) -- particularly protocol-specific data that needs to be conveyed all the way to the application -- then we should do this in a way so that the content can be trivially ignored by other protocols, and that other protocols can "add" their own protocol-specific data.

One good way to do this, IMO, is to locate options on the message using a key value pair.  It would also be possible to use dynamic IDs (sort of like some thread-specific-data schemes do), where we register a symbolic key, and get back a numeric ID, which might just become an array index.  (Although I suspect a simple key-value pair with a linked list is probably sufficient for most uses -- it might have a very minor impact on performance, but I'd think that walking a very short linked list of < 5 nodes would not do too much to seriously impact the performance.)

The details of how the data are attached to the message should be hidden behind an API however, so that we can adjust the details according to performance needs without breaking application consumers.